code/controllers/subsystem/bluespace_exploration.dm

```diff
@@ -0,0 +1,483 @@
+
+	//now select coordinates for a border turf
+	var/_x = rand(min,max)
+	var/_y = rand(min,max)
+
+	var/turf/T = locate(_x, _y, _z)
+	AM.forceMove(T)
+
+/datum/controller/subsystem/bluespace_exploration/proc/reset_turfs(list/turfs)
+	var/list/new_turfs = list()
+	for(var/i in turfs)
+		var/turf/T = i
+		var/turf/newT
+		if(istype(T, /turf/open/space))
+			newT = T
+		else
+			newT = T.ChangeTurf(/turf/open/space)
+		new_turfs += newT
+	return new_turfs
+
+//===================SPAWNING RUINS PROCS===================
+
+//TODO: Make this slower and spread over a time limit
+/datum/controller/subsystem/bluespace_exploration/proc/place_ruins(datum/data_holder/bluespace_exploration/data_holder)
+	//(Temp) get randomly created level
+	var/datum/exploration_location/location = new()
+	location.sector_features = list(FEATURE_ASTEROIDS)
+	//===Generate bluespace ruins===
+	var/list/bluespace_valid_ruins = list()
+	for(var/template_name in ruin_templates)
+		var/datum/map_template/ruin/exploration/ruin/R = ruin_templates[template_name]
+		var/valid = TRUE
+		for(var/feature in R.feature_type)
+			if(!(feature in location.sector_features))
+				valid = FALSE
+				break
+		if(!valid)
+			continue
+		bluespace_valid_ruins += R
+	//===Calculate standard ruins
+	var/datum/star_system/target_level = data_holder.target_star_system
+	var/list/standard_valid_ruins = list()
+	for(var/template_name in SSmapping.space_ruins_templates)
+		var/datum/map_template/ruin/space/R = SSmapping.space_ruins_templates[template_name]
+		standard_valid_ruins += R
+	//Generate Ruins
+	var/cost_limit = 20
+	while(cost_limit > 0)
+		if(!LAZYLEN(bluespace_valid_ruins))
+			break
+		var/list/selectable_ruins = list()
+		//TODO: Ruin weighting based on difficulty
+		//TODO: Bluespace ruins
+		if(target_level?.bluespace_ruins)
+			for(var/datum/map_template/ruin/exploration/ruin/R in bluespace_valid_ruins)
+				if(R.cost < cost_limit)
+					selectable_ruins += R
+		else
+			for(var/datum/map_template/ruin/space/R in standard_valid_ruins)
+				if(R.cost < cost_limit)
+					selectable_ruins += R
+		if(!LAZYLEN(selectable_ruins))
+			log_game("Ran out of selectable ruins, with [cost_limit] spawn points left.")
+			break
+		//Pick a ruin
+		var/datum/map_template/ruin/selected_ruin = pick(selectable_ruins)
+		if(!selected_ruin)
+			log_runtime("Warning, invalid ruin")
+			continue
+		if(target_level?.bluespace_ruins)
+			var/datum/map_template/ruin/exploration/ruin/BS_Ruin = selected_ruin
+			if(BS_Ruin.limited)
+				bluespace_valid_ruins -= BS_Ruin
+		else
+			var/datum/map_template/ruin/space/Space_ruin = selected_ruin
+			if(!Space_ruin.allow_duplicates)
+				standard_valid_ruins -= Space_ruin
+		//Subtract Cost
+		selected_ruin.try_to_place(data_holder.z_value, /area/space)
+		cost_limit -= selected_ruin.cost
+		CHECK_TICK
+	//=== Spawn Hostile Ships ===
+	var/max_ships = 3
+	var/threat_left = target_level.calculated_threat
+	while(threat_left > 0 && max_ships > 0)
```

Comment:
I feel like you should track a max number of attempts just to be safe.

Reply:
And also probably CHECK_TICK it.

Reply:
max_ships is the max number of attempts